pkg/datapath/linux/bigtcp/bigtcp.go

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bigtcp
+
+import (
+	datapathOption "github.com/cilium/cilium/pkg/datapath/option"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+const (
+	defaultGROMaxSize int = 65536
+	defaultGSOMaxSize int = 65536
+
+	BIGTCPGROMaxSize int = 196608
+	BIGTCPGSOMaxSize int = 196608
+)
+
+// Configuration is a BIG TCP configuration as returned by NewConfiguration
+type Configuration struct {
+	// gsoMaxSize is the GSO maximum size used when configuring devices
+	//
+	// Note that this is a singleton for the process including this
+	// package. This means, for instance, that when using this from the
+	// ``pkg/plugins/*`` sources, it will not respect the settings
+	// configured inside the ``daemon/``.
+	gsoMaxSize int
+
+	// groMaxSize is the GRO maximum size used when configuring devices
+	//
+	// Note that this is a singleton for the process including this
+	// package. This means, for instance, that when using this from the
+	// ``pkg/plugins/*`` sources, it will not respect the settings
+	// configured inside the ``daemon/``.
+	groMaxSize int
+}
+
+func (c *Configuration) GetGROMaxSize() int {
+	return c.groMaxSize
+}
+
+func (c *Configuration) GetGSOMaxSize() int {
+	return c.gsoMaxSize
+}
+
+func setGROGSOMaxSize(device string, GROMaxSize, GSOMaxSize int) error {
+	link, err := netlink.LinkByName(device)
+	if err != nil {
+		log.WithError(err).WithField("device", device).Warn("Link does not exist")
+		return nil
+	}
+
+	attrs := link.Attrs()
+	if (int(attrs.GROMaxSize) == GROMaxSize && int(attrs.GSOMaxSize) == GSOMaxSize) ||
+		(!option.Config.EnableIPv6BIGTCP &&
+			int(attrs.GROMaxSize) <= GROMaxSize &&
+			int(attrs.GSOMaxSize) <= GSOMaxSize) {
+		return nil
+	}
+
+	err = netlink.LinkSetGROMaxSize(link, GROMaxSize)
+	if err != nil {
+		return err
+	}
+
+	return netlink.LinkSetGSOMaxSize(link, GSOMaxSize)
+}
+
+func InitBIGTCP(bigtcpConfig *Configuration) {
+	var err error
+
+	if option.Config.DryMode {
+		return
+	}
+
+	if len(option.Config.GetDevices()) == 0 {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Warn("IPv6 BIG TCP could not detect host devices. Disabling the feature.")
+		}
+		option.Config.EnableIPv6BIGTCP = false
+		return
+	}
+
+	kernelGood := false
+	if h := probes.NewProbeManager().GetHelpers("sched_cls"); h != nil {
+		_, kernelGood = h["bpf_dynptr_data"]
+	}
+	if !kernelGood {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Fatalf("Cannot enable --%s, needs kernel 5.19 or newer",
```

Comment:
Do we really want to fatalf here? This seems extreme to me. In alpo for example we have a mix of kernel versions and we might upgrade some nodes to 5.19 when its available but this will stop us from enabling it in configMap. I would suggest if we really want to fatalf here we should have 'strict' and 'probe' modes.

Reply:
we can just disable the option with a log msg, but people must look for it to verify what happened, I think a hard error is better so they can't miss it, that's why it fatalfs in the requirement checks

Reply:
it will break deployment into our dogfooding cluster though.